File: 2024/C++/day3part2.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <string>

using namespace std;

int main() {
    ifstream file("../Input/day3.txt");
    if (!file) {
        cerr << "Error opening file!" << endl;
        return 1;
    }

    string line;
    int sum = 0;
    bool flag = true;
    regex pattern(R"((do\(\))|(don't\(\))|(mul\((\d+),(\d+)\)))");

    while(getline(file, line)) {
        sregex_iterator it(line.begin(), line.end(), pattern);
        sregex_iterator end;

        while(it != end) {
            smatch match = *it;

            if (!match[1].str().empty()) {
                flag = true;
            }
            if (!match[2].str().empty()) {
                flag = false;
            }
            if (!match[3].str().empty() && flag) { 
                int firstNumber = std::stoi(match[4].str());
                int secondNumber = std::stoi(match[5].str());
                sum += (firstNumber * secondNumber);
            }
            ++it;
        }
    }
    cout << sum << endl;
    return 0;
}

File: 2024/Python/day3part2.py
import re

sum = 0
pattern = r'do\(\)|don\'t\(\)|mul\(\d{1,3},\d{1,3}\)'
file = open("2024/Input/day3.txt")
flag = True
for line in file:
    line = re.findall(pattern, line)
    line = "".join(line)
    line = re.findall(r'do\(\)|don\'t\(\)|\(\d{1,3},\d{1,3}\)', line)
    for numbers in line:
        if numbers == "do()":
            flag = True
            continue
        if numbers == "don't()":
            flag = False
            continue
        if flag:
            numbers = numbers[1:len(numbers) - 1]
            firstNumber = numbers.split(',')[0]
            secondNumber = numbers.split(',')[1]
            sum = sum + (int(firstNumber) * int(secondNumber))
print(sum)